### PR TITLE
[gardening][lib/Basic][C++11] Replace `typedef` with `using`

### PR DESCRIPTION
**Pull request by ikesyo**
This is a series of pull requests attempting to replace `typedef`s with `using`s.

### CODE REVIEW

File: lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -82,16 +82,12 @@ llvm::StringRef swift::getProtocolName(KnownProtocolKind kind) {
...
 
 namespace {
-  typedef std::tuple<ClassDecl *, ObjCSelector, bool> ObjCMethodConflict;
+using ObjCMethodConflict = std::tuple<ClassDecl *, ObjCSelector, bool>;
 
-  /// An unsatisfied, optional @objc requirement in a protocol conformance.
-  typedef std::pair<DeclContext *, AbstractFunctionDecl *>
-    ObjCUnsatisfiedOptReq;
+/// An unsatisfied, optional @objc requirement in a protocol conformance.
+using ObjCUnsatisfiedOptReq = std::pair<DeclContext *, AbstractFunctionDecl *>;
```
Comment (ikesyo) [inline lib/AST/ASTContext.cpp]: The indentation changes are due to `git clang-format`.

**Comment by ikesyo** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
I guess `using` is often cleaner, but... what's the heuristic you're using to decide which typedefs to change?

**Comment by ikesyo** [comment]
I’m willing to replace most `typedef`s in subsequent PRs so there is no specific reasons for what I replaced in this PR. This is just a small start.

**Comment by ikesyo** [comment]
> but... what's the heuristic you're using to decide which typedefs to change?

Just narrowed the scope to only lib/Basic now.

@swift-ci smoke test

**Comment by ikesyo** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
Okay, thanks!

**Pull request merged**